connectors/sync_job_runner.py

```diff
@@ -214,8 +216,30 @@ async def _execute_content_sync_job(self, job_type, bulk_options):
             sync_rules_enabled=sync_rules_enabled,
             content_extraction_enabled=content_extraction_enabled,
             options=bulk_options,
+            optimize_on_meta_timestamp=self._lookup_service_config(
+                "enable_full_sync_meta_ts_optimization", False
+            ),
+            optimize_on_download_timestamp=self._lookup_service_config(
+                "enable_full_sync_download_ts_optimization", True
+            ),
         )
 
+    def _lookup_service_config(self, config_key, fallback):
+        """
+        Retrieve the source override config if present, otherwise return the top-level service config
+        :param config_key: the configuration to look up
+        :param fallback: the default value to use if the config_key is not found in the service config
+        :return: the configured value for config_key
+        """
+        default_value = self.service_config.get(config_key, fallback)
+        overrides = self.service_config.get("source_overrides", [])
```

Comment:
We spoke about it - we don't have source-specific config options now at all, and it's done on-purpose.

Content sources/connectors in current (and especially future) designs are supposed to be distributed. This change prevents this.

Our case is:

In current or serverless world, we don't have guarantee that a connector sync is executed on the same host. If hosts have different per-connector configs, then syncs will produce different results over time which is really hard to debug, support and grasp.

Config options like this should be done on the level of connector, IMO, and expressed in UI. Exposing them in config is a very slippery slope

Reply:
This was discussed at some length in our [project sync from Nov 9th](https://docs.google.com/document/d/1nV1CMdJVOrusbb53NoFrr4p0ee37To8wicyPfPOTCxs/edit#heading=h.4hda5l2bllxq). The issue is that a configuration through the UI is more involved, so we wanted to look at a solution that didn't require Kibana changes, and also didn't shoehorn this into an RCF that would need to be added individually to every connector.  We also discarded doing connector-specific YAML configs because that would require a process like:
1. create the connector in kibana
2. partially configure it
3. change the YAML for connectors to specify the optimization behavior
4. restart connectors (on Native, restart the whole node)
5. finish configuring and sync

which is not an ideal flow. The compromise we landed on was configuring by service type.

I expect in a multi-tenant serverless world we simply wouldn't expose this configuration - just like we wouldn't expose any `service` configs (like how many threads should be used).

Reply:
I agree we shouldn't expose any connector-specific or even source/service type-specific config in YAML, to make sync consistent. But this should only happen to cloud deployments of which we have control. 

For connector clients, developers can even customize the logic in whatever way they want, and I see no harm in giving the flexibility via config. If a developer runs a connector client, he/she should know what he/she is doing and be responsible for that.

Reply:
> For connector clients, developers can even customize the logic in whatever way they want, and I see no harm in giving the flexibility via config. If a developer runs a connector client, he/she should know what he/she is doing and be responsible for that.

That is true and we don't prevent forks, but on the level of framework this should not be supported by us - if anyone introduces unsupported logic then they are bound to support it, not us.

For me harm from flexibility via config (apart from problems with serverless) is also unnecessary complexity of connector setup.

So now you set up connector by:
1. First do setup via UI
2. Some features are behind a feature flag, so you go to dev tools and update setting there

If we allow customisation on config.yaml level, that's the third place. How do you now decide which config options go where? It's not very obvious to starrt.

But most importantly tp me it's just _too complex_ to manage -  maybe if you did your setup in a very simplistic way, everything works, you're happy. 

If the setup becomes more complex (you add redundancy, k8s, docker, tons of connector and instances running) then managing it becomes wildly more complex. Additionally, if something goes wrong in sync you'll have to dig into debugging much deeper. Let's say you have a problem with some sync that happened 4 days ago. You need to find out _who_ ran the sync, then if the host is still alive find it, then check its config file (if it was not edited along the way), etc, etc. It becomes a rabbit hole very fast and it's opposite of simplicity that we're trying to have.

So to me adding official support to config.yaml configuration per connector opens up tremendous complexity, which turns to slower development, much harder SDH and more potential errors/mistakes.

Reply:
The current idea is that we only support this for (customized) connector clients, which have fewer support demands than native connectors (or future multi-tenant connectors). To me this is just another way for users to customize their connector to fit their needs. 

It also fits where we want to go with connectors, which is to make them capable of running as standalone processes without managing their config through Elasticsearch. The advantages of this in the long run is that we can tell users concerned about secrets to store their secrets in the config.yml (populated through some other process probably).